src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -280,8 +280,15 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
                 .Build();
 
             yield return new FunctionOverloadBuilder("first")
-                // TODO even with non-literal types, some type arithmetic could be performed
-                .WithReturnResultBuilder(TryDeriveLiteralReturnType("first", LanguageConstants.Any), LanguageConstants.Any)
+                .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) =>
+                {
+                    return new(argumentTypes[0] switch 
+                    {
+                        TupleType tupleType => tupleType.Items.FirstOrDefault()?.Type ?? LanguageConstants.Null,
+                        ArrayType arrayType => TypeHelper.CreateTypeUnion(LanguageConstants.Null, arrayType.Item.Type),
+                        _ => throw new InvalidOperationException("Could not determine the return type from the provided input")
```

Comment:
```suggestion
                        _ => LanguageConstants.Any,
```

When I suggested the exception earlier, I didn't consider values of type `any`. Without this change, the following template would error since `first(untypedArray)` would be of type `any`:

```bicep
param untypedArray array

var a = first(first(untypedArray))
```